### PR TITLE
API: support sparse data frames

### PR DESCRIPTION
**Pull request by j-ittner**
- allow `OneHotEncoderDF` to produce sparse data frames
- accept pandas series as the `X` argument of all estimators, as an alternative to data frames
- add `CountVectorizerDF` and `TfidfVectorizerDF` (both of which expect a series or a single-column data frame as the `X` argument)
- preserve sparse dataframes in `ColumnTransformerDF` and `FeatureUnionDF`

### CODE REVIEW

File: src/sklearndf/pipeline/wrapper/_wrapper.py [hunk 1]
```diff
@@ -213,7 +219,25 @@ def _more_tags(self) -> Dict[str, Any]:
...
+    FeatureUnion transformer that returns sparse data frames instead of arrays if one or
+    more of its transformers return a sparse data frame.
+    """
+
+    # noinspection PyPep8Naming
+    def _hstack(
+        self, Xs: List[Union[npt.NDArray[Any], sparse.spmatrix, pd.DataFrame]]
+    ) -> Union[npt.NDArray[Any], sparse.spmatrix, pd.DataFrame]:
+        stacked = hstack_frames(Xs)
+        return stacked if stacked is None else super()._hstack(Xs)
```
Comment (mtsokol) [inline src/sklearndf/pipeline/wrapper/_wrapper.py]: I think `None` is missing here in the sum type that is returned. (As I understand `None` can be returned as `return stacked if stacked is None`)

Or shouldn't it be `return stacked if stacked is not None`?

File: src/sklearndf/wrapper/_wrapper.py [hunk 2]
```diff
@@ -448,14 +449,28 @@ def _post_fit(
...
+        # the expected columns.
+
+        if isinstance(X, pd.Series):
+            if X.name is None:
+                raise ValueError(
+                    "the name of the series passed as arg X must not be None"
+                )
+            X = X.to_frame()
+        elif not isinstance(X, pd.DataFrame):
             raise TypeError("arg X must be a DataFrame")
```
Comment (mtsokol) [inline src/sklearndf/wrapper/_wrapper.py]: Just a nitpick: `"arg X must be a DataFrame or a Series"`?

**Review by mtsokol — CHANGES_REQUESTED** [review]

File: src/sklearndf/pipeline/wrapper/_wrapper.py (hunk 1 above)
Comment (j-ittner) [reply src/sklearndf/pipeline/wrapper/_wrapper.py]: Yes it's a bug. Should be `if stacked is not None`. Will add a unit test.

File: src/sklearndf/wrapper/_wrapper.py (hunk 2 above)
Comment (j-ittner) [reply src/sklearndf/wrapper/_wrapper.py]: Indeed - it should be!

File: src/sklearndf/wrapper/_wrapper.py [hunk 3]
```diff
@@ -448,15 +449,29 @@ def _post_fit(
...
         self,
-        X: pd.DataFrame,
+        X: Union[pd.Series, pd.DataFrame],
         y: Optional[Union[pd.Series, pd.DataFrame]],
         *,
         expected_columns: pd.Index = None,
-    ) -> None:
-        if not isinstance(X, pd.DataFrame):
-            raise TypeError("arg X must be a DataFrame")
+    ) -> Tuple[pd.DataFrame, Union[pd.Series, pd.DataFrame]]:
```
Comment (mtsokol) [inline src/sklearndf/wrapper/_wrapper.py]: Just a nitpick - I think return type of this method for `y` part should be `Union[pd.Series, pd.DataFrame, None]` as `y` argument can be a `None` and is accepted.

**Review by mtsokol — APPROVED** [review]
Just one last comment - apart from that looks good to me!

File: src/sklearndf/wrapper/_wrapper.py (hunk 3 above)
Comment (j-ittner) [reply src/sklearndf/wrapper/_wrapper.py]: Well spotted, thanks. Fixed.

**Pull request merged**